plugins/by-name/typescript-tools/default.nix

```diff
@@ -157,38 +157,68 @@ in
       };
     };
   };
-  config = mkIf cfg.enable {
-    extraPlugins = [ cfg.package ];
-
-    plugins.lsp.postConfig =
-      with cfg;
-      let
-        options = {
-          inherit handlers;
-          on_attach = onAttach;
-
-          settings = with settings; {
-            separate_diagnostic_server = separateDiagnosticServer;
-            publish_diagnostic_on = publishDiagnosticOn;
-            expose_as_code_action = exposeAsCodeAction;
-            tsserver_path = tsserverPath;
-            tsserver_plugins = tsserverPlugins;
-            tsserver_max_memory = tsserverMaxMemory;
-            tsserver_format_options = tsserverFormatOptions;
-            tsserver_file_preferences = tsserverFilePreferences;
-            tsserver_locale = tsserverLocale;
-            complete_function_calls = completeFunctionCalls;
-            include_completions_with_insert_text = includeCompletionsWithInsertText;
-            code_lens = codeLens;
-            disable_member_code_lens = disableMemberCodeLens;
-            jsx_close_tag = with jsxCloseTag; {
-              inherit enable filetypes;
-            };
+
+  # NOTE: call setup in lsp config
+  callSetup = false;
+  extraConfig =
+    cfg:
+    let
+      definedOpts = lib.filter (opt: lib.hasAttrByPath opt cfg.settings) [
+        [ "separateDiagnosticServer" ]
+        [ "publishDiagnosticOn" ]
+        [ "exposeAsCodeAction" ]
+        [ "tsserverPath" ]
+        [ "tsserverPlugins" ]
+        [ "tsserverMaxMemory" ]
+        [ "tsserverFormatOptions" ]
+        [ "tsserverFilePreferences" ]
+        [ "tsserverLocale" ]
+        [ "completeFunctionCalls" ]
+        [ "includeCompletionsWithInsertText" ]
+        [ "codeLens" ]
+        [ "disableMemberCodeLens" ]
+        [
+          "jsxCloseTag"
+          "enable"
+        ]
+        [
+          "jsxCloseTag"
+          "filetypes"
+        ]
+      ];
+    in
+    {
+      plugins.lsp.postConfig =
+        let
+          # TODO:: introduced 10-22-2024: remove after 24.11
+          renamedSettings = lib.foldl' (
+            acc: optPath:
+            let
+              snakeCaseName = lib.concatStringsSep "_" (lib.map lib.nixvim.toSnakeCase (lib.toList optPath));
```

Comment:
I don't like this at all, but I keep breaking it when I change it...

Reply:
This doesn't preserve the nested attrs. E.g. `jsxCloseTag.enable` will be renamed to `jsx_close_tag_enable`

Instead we'll need to replace `snakeCaseName` with a `snakeCasePath`, and combine [`recursiveUpdate`](https://noogle.dev/f/lib/recursiveUpdate) with [`setAttrByPath`](https://noogle.dev/f/lib/setAttrByPath).

We also don't really need to handle the null-checking here, since `toLuaObject` will do that for us. Even though I wish it didn't...

We can use [`getAttrFromPath`](https://noogle.dev/f/lib/getAttrFromPath) with no fallback value, because `definedOpts` has already been filtered using [`hasAttrByPath`](https://noogle.dev/f/lib/hasAttrByPath).

We don't need `toList` either, because you already normalized `definedOpts` to all be attrpath-lists.

```nix
renamedSettings = lib.foldl' (
  acc: optPath:
  let
    snakeCasePath = builtins.map lib.nixvim.toSnakeCase optPath;
    optValue = lib.getAttrFromPath optPath cfg.settings;
  in
  lib.recursiveUpdate acc (lib.setAttrByPath snakeCasePath optValue)
) { } definedOpts;
```